### PR TITLE
Finalize target block after warp sync

### PR DESCRIPTION
**Pull request by davxy**
## TL;DR

This PR addresses the following issues:
1. Level-monitor `restore` procedure after warp sync and if node is restarted during block-history download
2. Finalization of target block after warp sync
3. Proper handling of finalized blocks during block history download
4. prevent insertion of finalized blocks in level-monitor (check if block > final)

Fixes: https://github.com/paritytech/cumulus/issues/2659

May help with: https://github.com/paritytech/substrate/issues/13202

Depends on https://github.com/paritytech/substrate/pull/14308

---

## Warping without the fix

Finalized block number is not updated until we've not finished block history download.

This is a source of issues in case we stop and restart the node while we are downloading block history.

Issues:
1. Level monitor panics during `restore` procedure: we are trying to compute a route from last finalized (that is still 0) to each of the leaves. Because the blocks in the gap were not downloaded yet, this triggers a panic.
2. If we don't stop the node during block-history download then download/import of the gap can be memory-expensive since each of the blocks imported (actually the hash) will be inserted within the monitor (number > finalized = 0)
   The monitor will end up containing the full block history from gap start to the leaves.

At the end of the block-history download finalized flag is finally set to the proper value, thus the monitor is cleaned up.

We can do a lot better and mimic what the relay chain actually does (see "Warping with fix below").

**Log before fix (statemine)**

```rust
⏩ Warping, Downloading state, 2.34 Mib (9 peers), best: #0 (0x4823…771a), finalized #0 (0x4823…771a), ⬇ 343.0kiB/s ⬆ 22.5kiB/s
⏩ Warping, Downloading state, 7.18 Mib (10 peers), best: #0 (0x4823…771a), finalized #0 (0x4823…771a), ⬇ 625.9kiB/s ⬆ 2.7kiB/s
⏩ Warping, Downloading state, 11.87 Mib (10 peers), best: #0 (0x4823…771a), finalized #0 (0x4823…771a), ⬇ 842.7kiB/s ⬆ 10.1kiB/s
⏩ Warping, Downloading state, 16.73 Mib (10 peers), best: #0 (0x4823…771a), finalized #0 (0x4823…771a), ⬇ 399.1kiB/s ⬆ 1.3kiB/s
⏩ Warping, Downloading state, 23.15 Mib (10 peers), best: #0 (0x4823…771a), finalized #0 (0x4823…771a), ⬇ 576.1kiB/s ⬆ 0.2kiB/s
⏩ Warping, Downloading state, 31.52 Mib (10 peers), best: #0 (0x4823…771a), finalized #0 (0x4823…771a), ⬇ 788.7kiB/s ⬆ 0.7kiB/s

⏩ Warping, Importing state, 34.97 Mib (10 peers), best: #0 (0x4823…771a), finalized #0 (0x4823…771a), ⬇ 337.7kiB/s ⬆ 1.3kiB/s
⏩ Warping, Importing state, 34.97 Mib (10 peers), best: #0 (0x4823…771a), finalized #0 (0x4823…771a), ⬇ 0.2kiB/s ⬆ 0.2kiB/s
⏩ Warping, Importing state, 34.97 Mib (10 peers), best: #0 (0x4823…771a), finalized #0 (0x4823…771a), ⬇ 0.7kiB/s ⬆ 0.7kiB/s
⏩ Warping, Importing state, 34.97 Mib (10 peers), best: #0 (0x4823…771a), finalized #0 (0x4823…771a), ⬇ 1.3kiB/s ⬆ 1.1kiB/s

Warp sync is complete (34 MiB), restarting block sync.

⏩ Block history, #3009 (2 peers), best: #4642329 (0x5467…2813), finalized #0 (0x4823…771a), ⬇ 385.4kiB/s ⬆ 2.6kiB/s
⏩ Block history, #3265 (2 peers), best: #4642329 (0x5467…2813), finalized #0 (0x4823…771a), ⬇ 193.8kiB/s ⬆ 0.6kiB/s
⏩ Block history, #7425 (2 peers), best: #4642329 (0x5467…2813), finalized #0 (0x4823…771a), ⬇ 741.0kiB/s ⬆ 1.1kiB/s
...
⏩ Block history, #4639041 (23 peers), best: #4643003 (0x3d91…982f), finalized #0 (0x4823…771a), ⬇ 5.7MiB/s ⬆ 1.3kiB/s
⏩ Block history, #4639681 (23 peers), best: #4643003 (0x3d91…982f), finalized #0 (0x4823…771a), ⬇ 1.6MiB/s ⬆ 0.3kiB/s
⏩ Block history, #4639681 (23 peers), best: #4643004 (0x4b70…d135), finalized #0 (0x4823…771a), ⬇ 157.3kiB/s ⬆ 2.1kiB/s
⏩ Block history, #4639873 (23 peers), best: #4643004 (0x4b70…d135), finalized #0 (0x4823…771a), ⬇ 665.1kiB/s ⬆ 0.2kiB/s
⏩ Block history, #4642305 (22 peers), best: #4643005 (0x2c42…bca4), finalized #0 (0x4823…771a), ⬇ 1.5MiB/s ⬆ 10.6kiB/s

Block history download is complete.

💤 Idle (23 peers), best: #4643064 (0xdf89…2480), finalized #4643014 (0xba26…234a), ⬇ 313.4kiB/s ⬆ 5.0kiB/s
💤 Idle (23 peers), best: #4643064 (0xdf89…2480), finalized #4643062 (0xccf9…323a), ⬇ 64 B/s ⬆ 64 B/s
💤 Idle (22 peers), best: #4643065 (0x498a…6f9c), finalized #4643063 (0x1386…8998), ⬇ 47.5kiB/s ⬆ 27.6kiB/s
...
```

## Warping with the fix

Once we discovered what is the last target block (according to the relay-chain) we should import this block as final (i.e. setting the `BlockImportParams::finalized` flag to `true`).

The adopted strategy mimics what `grandpa` already does [here](https://github.com/paritytech/substrate/blob/13ed4508e65992b761e4a713bd832aa38cc08ec7/client/consensus/grandpa/src/import.rs#L544-L546) and thus [here](https://github.com/paritytech/substrate/blob/13ed4508e65992b761e4a713bd832aa38cc08ec7/client/consensus/grandpa/src/import.rs#L467).

That is, **the finalized flag is set to true if the block also contains the state**.
AFAIK in practice this is true only for the warp sync target block.

**Log after the fix (statemine)**:

```rust
⏩ Warping, Downloading state, 7.18 Mib (8 peers), best: #0 (0x4823…771a), finalized #0 (0x4823…771a), ⬇ 962.2kiB/s ⬆ 22.7kiB/s
⏩ Warping, Downloading state, 9.59 Mib (8 peers), best: #0 (0x4823…771a), finalized #0 (0x4823…771a), ⬇ 320.5kiB/s ⬆ 0.3kiB/s
⏩ Warping, Downloading state, 23.15 Mib (8 peers), best: #0 (0x4823…771a), finalized #0 (0x4823…771a), ⬇ 1.4MiB/s ⬆ 1.8kiB/s
⏩ Warping, Downloading state, 31.52 Mib (8 peers), best: #0 (0x4823…771a), finalized #0 (0x4823…771a), ⬇ 821.4kiB/s ⬆ 1.4kiB/s
⏩ Warping, Downloading state, 31.52 Mib (8 peers), best: #0 (0x4823…771a), finalized #0 (0x4823…771a), ⬇ 185.7kiB/s ⬆ 0.1kiB/s

⏩ Warping, Importing state, 34.97 Mib (8 peers), best: #0 (0x4823…771a), finalized #0 (0x4823…771a), ⬇ 128.3kiB/s ⬆ 0.2kiB/s
⏩ Warping, Importing state, 34.97 Mib (8 peers), best: #0 (0x4823…771a), finalized #0 (0x4823…771a), ⬇ 1.1kiB/s ⬆ 1.1kiB/s
⏩ Warping, Importing state, 34.97 Mib (8 peers), best: #0 (0x4823…771a), finalized #0 (0x4823…771a), ⬇ 0.2kiB/s ⬆ 0.2kiB/s
⏩ Warping, Importing state, 34.97 Mib (8 peers), best: #0 (0x4823…771a), finalized #0 (0x4823…771a), ⬇ 0.2kiB/s ⬆ 0.2kiB/s
⏩ Warping, Importing state, 34.97 Mib (8 peers), best: #0 (0x4823…771a), finalized #0 (0x4823…771a), ⬇ 1.0kiB/s ⬆ 1.1kiB/s

Warp sync is complete (34 MiB), restarting block sync.

⏩ Block history, #2112 (7 peers), best: #4641676 (0x2e24…a3db), finalized #4641676 (0x2e24…a3db), ⬇ 669.7kiB/s ⬆ 1.2kiB/s    
⏩ Block history, #2112 (1 peers), best: #4641678 (0xfad2…3364), finalized #4641678 (0xfad2…3364), ⬇ 53.7kiB/s ⬆ 2.2kiB/s    
⏩ Block history, #2176 (2 peers), best: #4641683 (0x4692…8354), finalized #4641681 (0x1a3d…52fa), ⬇ 95.0kiB/s ⬆ 13.9kiB/s    
...
⏩ Block history, #4635712 (17 peers), best: #4642215 (0x60de…a865), finalized #4642214 (0x3d8e…eebe), ⬇ 4.5MiB/s ⬆ 0.7kiB/s
⏩ Block history, #4638848 (17 peers), best: #4642216 (0x770b…69bc), finalized #4642214 (0x3d8e…eebe), ⬇ 7.4MiB/s ⬆ 1.5kiB/s
⏩ Block history, #4641675 (17 peers), best: #4642216 (0x770b…69bc), finalized #4642214 (0x3d8e…eebe), ⬇ 6.2MiB/s ⬆ 22.7kiB/s

Block history download is complete.

💤 Idle (17 peers), best: #4642216 (0x770b…69bc), finalized #4642215 (0x60de…a865), ⬇ 3.3kiB/s ⬆ 1.9kiB/s
💤 Idle (17 peers), best: #4642217 (0xf995…9fec), finalized #4642215 (0x60de…a865), ⬇ 1.6kiB/s ⬆ 1.5kiB/s
💤 Idle (17 peers), best: #4642218 (0x8044…8b93), finalized #4642216 (0x770b…69bc), ⬇ 56 B/s ⬆ 51 B/s
...
```

As can be seen the finalized block correctly starts being updated after state is imported, as expected.

---

## Additional notes

Cumulus node caches the parachain blocks that are signaled as final [here](https://github.com/paritytech/cumulus/blob/5b0db49b5c2ed417d3673ecc80910cf92f103bf1/client/consensus/common/src/parachain_consensus.rs#L141)
and triggers a `Finalized::finalize_block()` for each block found in the cache.

This procedure marks the block as final BUT doesn't check if the block that we are finalizing is below the current "best" block.

As a consequence may happen that the finalized block number > best. 
We don't want this as it also by-passes some general assumptions spread around the code.

To correctly handle cases like this a more defensive check been introduced in Substrate by https://github.com/paritytech/substrate/pull/14308

### CODE REVIEW

File: client/consensus/common/src/lib.rs [hunk 1]
```diff
@@ -147,6 +147,11 @@ where
 		let hash = params.post_hash();
 		let number = *params.header.number();
 
+		if params.with_state() {
+			// Force imported state finality
```
Comment (skunert) [inline client/consensus/common/src/lib.rs:151]:
```suggestion
			// Force imported state finality.
			// Required for warp sync. We assume that preconditions have been
			// checked properly and we are importing a finalized block with state.
```
Nit: Maybe a bit more descriptive, feel free to ignore.

**Review by skunert — APPROVED** [review]

File: client/consensus/common/src/lib.rs [hunk 2]
```diff
@@ -147,6 +147,11 @@ where
...
+			params.finalized = true;
```
Comment (bkchr) [inline client/consensus/common/src/lib.rs:152]: It would be nice to have some tests for it, but I also don't have any good idea where to put it.

Why did you decided against the approach of doing this in `WarpSyncEngine` in Substrate?

**Review by bkchr — APPROVED** [review]
